pep-9999.rst

```diff
@@ -0,0 +1,284 @@
+    cert                str          Path to certificate authority bundle (*.pem)
+    cert-client         str          Path to client certificate (*.pem)
+    enabled             bool         Whether the repository should be used or not (default: true)
+    package_index_name  str          The corresponding package index name, defined in ``config.toml``
+    password            str          The password for the repository
+    proxy_password      str          The password for the proxy server
+    proxy_username      str          The username for the proxy server
+    username            str          The username for the repository
+    ==================  ===========  ===============================================
+
+
+Tools preferring to look up credentials from other locations, such as keyring, are
+advised to use the full package index URL as the key, rather than the index name or
+hostname.
+
+
+The auth.toml file
+''''''''''''''''''
+
+Optionally, authentication can be defined in a separate ``auth.toml``. This way,
+credentials can be fully or partially separated from package index URLs.
+
+Example ``auth.toml``::
+
+    [[credentials]]
+    package_index_name = "myrepo1"
+    username = "foo"
+    password = "YmFyCg=="
+
+
+Package discovery
+-----------------
+
+PyPI.org
+''''''''
+
+The PyPI.org repository (``repositories.pypi``) does not have to be specified in the
+``config.toml``. If not specified, it will be used as the first package index by
+default. However, by explicitly defining it, the repository attributes can be used and
+it can be prioritized differently.
+
+
+Package index priority
+''''''''''''''''''''''
+
+Tools should respect priority by acting as if all lower priority package index were not
+specified when a higher priority package index satisfies the task. Tools that choose not
+to respect priority MUST display a warning to users when a non-zero priority value is
+being ignored. It is acceptable for tools to be less efficient in order to handle
+prioritized package indexes, for example, by not performing queries in parallel.
+
+
+Configuration file discovery
+----------------------------
+
+The locations of the configuration files (``config.toml`` and ``auth.toml``) vary slightly across platforms. You may have global, per-user and per-environment configuration:
+
+If multiple configuration files are found, then they are merged in the following order:
+
+1. The global files are read
+2. The per-user files are read
+3. The per-environment files are read
+
+
+Global
+''''''
+
+* Unix: ``/etc/.config/python/``
+* macOS: ``/Library/Application Support/python/``
+* Windows: ``%PROGRAMDATA%\Python``
+
+
+Per-user
+''''''''
+
+* Unix: ``$HOME/.config/python/``
+* macOS: ``$HOME/.config/python/``
+* Windows: ``%APPDATA%\Python\`` (note, this is Roaming app data, not Local)
+
+
+Per-environment
+'''''''''''''''
+
+* Unix: ``sys.prefix``
+* macOS: ``sys.prefix``
+* Windows: ``sys.prefix``
+
+
```

Comment:
@abn @zooba what do you think about simply specifying the indexes/credentials per-project?

```suggestion
Per-project
'''''''''''
``pyproject.toml``
```

Reply:
I don't think credentials should be here. Index configuration sure. But that just looks at formalising what poetry defines as sources right?

Reply:
I think anywhere you specify an index URL, you should be able to specify the credentials needed to access that URL.

Should you _always_ specify them there? Probably not, security risks and all. But there's no reason to forbid it. Credentials are often included in the URL (e.g. `https://user:pass@site.com`).

Should you be able to specify/override them in another config file? Yes. Which means if you specify an index name in multiple files, they'll merge together in some predictable manner.

Then if the auth is only in the project's TOML file (or one that's explicitly passed to the tool through a CLI option or environment variable), it'll merge in just the same as if it were specified anywhere else.

In other words, there's one complete schema of options, and one set of rules for merging multiple TOML files to fill out that single set of options. (Then ideally a shared library to do that merge for everyone.)